File: Quote.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace stockmarket
{
    internal class Quote
    {
        internal int month { get; private set; }
        internal int day { get; private set; }
        internal int year { get; private set; }
        internal decimal open { get; private set; }
        internal decimal close { get; private set; }
        internal decimal high { get; private set; }
        internal decimal low { get; private set; }
        internal long volume { get; private set; }
        internal Quote(int m, int d, int y,
            decimal o, decimal c, decimal h, decimal l,
            long v)
        {
            month = m;
            day = d;
            year = y;
            open = o;
            close = c;
            high = h;
            low = l;
            volume = v;
        }
    }
}

File: Quote.hs
module Quote where

import Data.Int
import qualified Test.QuickCheck as QC

data Quote = Quote {symbol :: String,
                    day :: String,
                    index :: Int,
                    volume :: Int64,
                    open, high, low, close :: Double}
            deriving (Show)

instance QC.Arbitrary Quote where
    arbitrary = do
        a <- QC.arbitrary
        b <- QC.arbitrary
        c <- QC.arbitrary
        d <- QC.arbitrary
        e <- QC.arbitrary
        f <- QC.arbitrary
        g <- QC.arbitrary
        h <- QC.arbitrary
        return (Quote a b (abs c) (abs d) (abs e) (abs f) (abs g) (abs h))
        
makeQuote :: Int -> [String] -> Quote
makeQuote idx (a:b:c:d:e:f:g:h:[]) =    let ratio = (read h::Double) / (read f::Double)
                                            adjOpen = (read c::Double) * ratio
                                            adjHigh = (read d::Double) * ratio
                                            adjLow = (read e::Double) * ratio
                                            adjClose = read h::Double -- also e * ratio
                                            vol = read g::Int64
                                        in  Quote a b idx vol adjOpen adjHigh adjLow adjClose
makeQuote _ _ = error "Insufficient data to make a quote"

makeQuotes :: [[String]] -> [Quote]
makeQuotes = makeQuotes' 0

makeQuotes' :: Int -> [[String]] -> [Quote]
makeQuotes' idx (x:xs) = (makeQuote idx x):(makeQuotes' (idx + 1) xs)
makeQuotes' _ [] = []
